### PR TITLE
Add `#[must_not_suspend]` to some types in std

### PR DESCRIPTION
**Pull request by guswynn**
I am not sure what else should have it? `Ref`?

### CODE REVIEW

File: library/std/src/sync/mutex.rs [hunk 1]
```diff
@@ -188,6 +188,12 @@ unsafe impl<T: ?Sized + Send> Sync for Mutex<T> {}
 /// [`lock`]: Mutex::lock
 /// [`try_lock`]: Mutex::try_lock
 #[must_use = "if unused the Mutex will immediately unlock"]
+#[cfg_attr(
+    not(bootstrap),
+    must_not_suspend = "Holding a MutexGuard across suspend \
+                      points can cause deadlocks, delays, \
+                      and cause Future's to not implement `Send`"
```
Comment (Boddlnagg) [inline library/std/src/sync/mutex.rs:195]: A plural-s does not go with an apostrophe (same issue `rwlock.rs`).

**Comment by Kobzol** [comment]
I would personally really like to see it on `Ref` and `RefMut`. When single-threaded async executors are used with Refcells, holding a `Ref` across `await` is a pretty common source of `BorrowMut` errors in my experience. There's a corresponding `clippy` [lint](https://rust-lang.github.io/rust-clippy/master/#await_holding_refcell_ref) for this.

**Review by dtolnay — APPROVED** [review]
The RFC explicitly calls out `MutexGuard` as a motivation for the lint, so that one is a no-brainer. The other types here are all very clearly in the same category as `MutexGuard` &mdash; LGTM.

**Comment by dtolnay** [comment]
Cross linking tracking issue. https://github.com/rust-lang/rust/issues/83310

**Pull request merged**